src/Build.UnitTests/BackEnd/IntrinsicTask_Tests.cs

```diff
@@ -1714,6 +1713,179 @@ public void RemoveAllItemsInList()
+        }
+
+        [Fact]
+        public void RemoveWithItemReferenceOnMatchingMetadataReferences()
+        {
+            // <PropertyGroup>
+            //     <p0>v0</p0>
+            // </PropertyGroup>
+            string content = ObjectModelHelpers.CleanupFileContents(
+                @"<Project ToolsVersion='msbuilddefaulttoolsversion' xmlns='msbuildnamespace'>
+                    <Target Name='t'>
+                        <ItemGroup>
+                            <Meta2 Include='M2'/>
+                            <Meta2 Include='g'/>
+
+                            <I1 Include='a1' v0='1' M2='a'/>
+                            <I1 Include='b1' v0='2' M2='x'/>
+                            <I1 Include='c1' v0='3' M2='y'/>
+                            <I1 Include='d1' v0='4' M2='b'/>
+
+                            <I2 Include='a2' v0='x' m2='c'/>
+                            <I2 Include='b2' v0='2' m2='x'/>
+                            <I2 Include='c2' v0='3' m2='Y'/>
+                            <I2 Include='d2' v0='y' m2='d'/>
+
+                            <I2 Remove='@(I1)' MatchOnMetadata='$(p0);@(Meta2)' />
+                        </ItemGroup>
+                    </Target></Project>");
+
+            IntrinsicTask task = CreateIntrinsicTask(content);
+            PropertyDictionary<ProjectPropertyInstance> properties = GeneratePropertyGroup();
+            Lookup lookup = LookupHelpers.CreateLookup(properties);
+            ExecuteTask(task, lookup);
+
+            var items = lookup.GetItems("I2");
+
+            items.Select(i => i.EvaluatedInclude).ShouldBe(new[] { "a2", "c2", "d2" });
+
+            items.ElementAt(0).GetMetadataValue("v0").ShouldBe("x");
+            items.ElementAt(0).GetMetadataValue("M2").ShouldBe("c");
+            items.ElementAt(1).GetMetadataValue("v0").ShouldBe("3");
+            items.ElementAt(1).GetMetadataValue("M2").ShouldBe("Y");
+            items.ElementAt(2).GetMetadataValue("v0").ShouldBe("y");
+            items.ElementAt(2).GetMetadataValue("M2").ShouldBe("d");
+        }
+
+        [Fact]
+        public void KeepWithItemReferenceOnNonmatchingMetadata()
+        {
+            string content = ObjectModelHelpers.CleanupFileContents(
+                @"<Project ToolsVersion='msbuilddefaulttoolsversion' xmlns='msbuildnamespace'>
+                    <Target Name='t'>
+                        <ItemGroup>
+                            <I1 Include='a1' a='1' b='a'/>
+                            <I1 Include='b1' a='2' b='x'/>
+                            <I1 Include='c1' a='3' b='y'/>
+                            <I1 Include='d1' a='4' b='b'/>
+
+                            <I2 Include='a2' c='x' d='c'/>
+                            <I2 Include='b2' c='2' d='x'/>
+                            <I2 Include='c2' c='3' d='Y'/>
+                            <I2 Include='d2' c='y' d='d'/>
+
+                            <I2 Remove='@(I1)' MatchOnMetadata='e;f;g' />
```

Comment:
If [nonexisting metadata should not match](https://github.com/microsoft/msbuild/pull/4997#discussion_r363516251), then this test is broken right? It should match everything and remove all items.

It does kind of look strange to remove all items here. It might make MatchOnMetadata more error prone w.r.t. to users typing in msbuild code with unintended consequences. If the user has two items out of which only a subset define the metadata to match on, it might kind of be weird / unexpected to also remove the subset that does not define the metadata.

Reply:
Right now I have it so that there has to be at least one nontrivial match to remove. That is, if I1 and I2 both do not define any of the metadata specified, then nothing is removed. 
It would probably make sense to make one, consistent call here. I don't feel equipped to give input as to which version makes more sense.

Reply:
If you want to make it more complicated (but perhaps more functional?), you could add a flag such that if it's set, you remove everything that doesn't disagree with any of the metadata provided, and if it isn't set, you remove only things that agree with every bit of metadata. That's probably too complicated to be worth it, but I thought I should mention it.

Reply:
Ideally we should come up with a reasonable default, but if necessary we could come up with an additional `MatchOnMetadataOptions` metadata.

Let's discuss it in the next PR review, or a separate meeting, however you wish.